runtime/doc/lsp.txt

```diff
@@ -1689,21 +1689,25 @@ code_action({opts})                                *vim.lsp.buf.code_action()*
       • https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_codeAction
       • vim.lsp.protocol.CodeActionTriggerKind
 
-declaration({opts})                                *vim.lsp.buf.declaration()*
+declaration({opts}, {callback})                    *vim.lsp.buf.declaration()*
```

Comment:
is there an `on_foo` name we can use? same for the others. 
```suggestion
declaration({opts}, {on_result})                    *vim.lsp.buf.declaration()*
```

Reply:
I think `callback` feels better as `result` does have a "positive" connotation, but this callback should always be invoked regardless of the request completing successfully or not.

Reply:
All of the callbacks in this pr need some sort of `on_` prefix (`:help dev-naming`). We use "callback" way too much.

Reply:
`callback` is usually a continuation and is the async version of `return`. `on_` isn't necessary for continuations.

Reply:
> leads lots of variations of the same name: `on_exit`, `on_done`, `on_finished`, `on_result`, `on_complet`

If those are the same then they should align on one name.

> It's also not obvious these names are callbacks... unlike `callback`.

Anything prefixed with `on_` is a callback.

> but the `on_x` thing just seems like a specific thing you are over pushing for everything and seems to come directly from js.

No, this decision was made after experience from early days of the project where `_cb` prefix was used. `on_` is much more readable and conventional. 

It's a project guideline, and objecting to project guidelines continually is really not helpful.

Reply:
> A continuation is just a special case. And naming it differently leads to confusion and debates, which is exactly what I'm trying to avoid.

Arguably it's the most common case.

Reply:
For what's worth, +1 for `on_` from my side.

There's already:

```
                                                        *dev-name-events*
Use the "on_" prefix to name event-handling callbacks and also the interface for
"registering" such handlers (on_key). The dual nature is acceptable to avoid
a confused collection of naming conventions for these related concepts.
```

This fits here as well, not sure why we'd special case continuation cases.

Reply:
> This fits here as well, not sure why we'd special case continuation cases.

`callback` fits better here, and does in most places and is a much more common and understood name. There's a reason people use it, and I don't think it's because we already use `callback` for autocmds. I've been reading about a lot of different structured concurrency frameworks and they all seem to use this, except js which coincidentally seems to have one the worst frameworks.

Reply:
I added it because we had no guidelines and we have things named `handler`, `callback`, `x_cb`, `on_x`, and more. (btw, js uses "listener")

How about this compromise: 

- `callback` for continuations (exactly one callback, not multiple like [uv.new_work](https://github.com/neovim/neovim/blob/4745270bf124a960ccdffdddbb6c27b7bf36ba82/runtime/doc/luvref.txt#L3600) for example), 
- `on_x` for everything else

and we can commit to this (no debates about special cases or personal preferences in the future)? Then I will update the docs.

Reply:
sgtm.

For the `new_work` case, `after_work_callback` is the continuation and `work_callback` is the work and is quite special because it runs with its own Lua state. So you could have `work` and `callback` as arguments for it to follow the rule. We don't own those anyway, and fwiw I would accept `on_` there if we could think of appropriate names.

Reply:
> callback fits better here, and does in most places and is a much more common and understood name. There's a reason people use it, and I don't think it's because we already use callback for autocmds

Conclusory language?
Yes, callback is useful to describe the concept of a higher-order function used as parameter to signal something back to the original caller. `on_` naming is useful to clarify _when_ or _what_ it is that is signaled. E.g. I'd still describe a `on_list` as a callback, but that doesn't mean that the parameter needs to be named `callback`

> callback for continuations

Isn't that to some degree up to the call-site if a callback is used for continuation or not?